torch/nested/_internal/ops.py

```diff
@@ -255,19 +267,461 @@ def jagged_binary_pointwise(func, *args, **kwargs):
+            f"and value.device: {value.device} instead."
+        )
+    if query.dim() < 2 or key.dim() < 2 or value.dim() < 2:
+        raise ValueError(
+            f"Expected query, key, and value to all be  at least 2 dimensional, but got query.dim: "
+            f"{query.dim()}, key.dim: {key.dim()} and value.dim: {value.dim()} instead."
+        )
+    if query._ragged_idx != 2 or key._ragged_idx != 2 or value._ragged_idx != 2:
+        raise ValueError(
+            f"Expected query, key, and value to all be be jagged at dimension 2, but got query._ragged_idx: "
+            f"{query._ragged_idx}, key._ragged_idx: {key._ragged_idx} and value._ragged_idx: {value._ragged_idx} instead."
+        )
+    if attn_mask is not None:
+        # TODO: Figure out whether masks are actually supported for this layout or not
+        if attn_mask.dtype != torch.bool and attn_mask.dtype != query.dtype:
+            raise ValueError(
+                f"Expected attn_mask dtype to be bool or to match query dtype, but got attn_mask.dtype: "
+                f"{attn_mask.dtype}, and query.dtype: {query.dtype} instead."
+            )
+
+
+def _can_use_math_sdpa_jagged(
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    attn_mask: Optional[torch.Tensor] = None,
+    dropout_p=0.0,
+    is_causal=False,
+    scale=None,
+):
+    if (
+        not query.is_contiguous()
+        or not key.is_contiguous()
+        or not value.is_contiguous()
+    ):
+        raise ValueError("If inputs are nested tensors they must be contiguous.")
+    if is_causal:
+        raise ValueError(
+            "Nested tensors for query / key are not supported when is_causal=True."
+        )
+
+
+def _select_sdp_backend(query, key, value, attn_mask, dropout, is_causal):
+    if (
+        not flash_sdp_enabled()
+        and not mem_efficient_sdp_enabled()
+        and not math_sdp_enabled()
+    ):
+        return SDPBackend.ERROR
+
+    ordering = (
+        SDPBackend.FLASH_ATTENTION,
+        SDPBackend.EFFICIENT_ATTENTION,
+        SDPBackend.MATH,
+    )
+
+    params = SDPAParams(query, key, value, attn_mask, dropout, is_causal)
+
+    for backend in ordering:
+        if backend == SDPBackend.FLASH_ATTENTION:
+            if can_use_flash_attention(params):
+                return SDPBackend.FLASH_ATTENTION
+        if backend == SDPBackend.EFFICIENT_ATTENTION:
+            if can_use_efficient_attention(params):
+                return SDPBackend.EFFICIENT_ATTENTION
+        if backend == SDPBackend.EFFICIENT_ATTENTION:
+            if math_sdp_enabled():
+                return SDPBackend.MATH
+
+    log.warning("Memory efficient kernel not used because:")
+    can_use_efficient_attention(params, True)
+    log.warning("Flash attention kernel not used because:")
+    can_use_flash_attention(params, True)
+    raise ValueError("No available kernel. Aborting execution.")
```

Comment:
We return `SDPBackend.ERROR` above but error here - should we also return `SDPBackend.ERROR` here?

Reply:
if we raise there is no chance for the function to return right? In C++, this function also raises with an always failing TORCH_CHECK()

Reply:
@drisspg what's `SDPBackend.ERROR` for? feels like we should signal consistently from this function (either ValueError or return this, leaning towards the former). The two error cases I see:
* No kernels are "enabled"
* None of the "enabled" kernels are available

Reply:
I like this approach more, changing to return ERROR and centralizing errors

Reply:
Error was is the return type of select_sdp_backend if no valid kernel is found.. that being said we will always error out with a "TORCH_CHECK" prior to actually returning from this function so in practice it is not really used but more of a belt and suspenders type of deal.